### PR TITLE
Removing more generics for #392

### PR DESCRIPTION
**Pull request by alexeyzimarev**
I chose a bit different approach compared to Observer metrics because some of the methods are actually not bound to `long` and `double`. So, some generics remain but all if statements to check if the type is double or long are removed.

I am a bit concerned about the overall result, because the `Collect` method doesn't seem to have tests, but it should work.

### CODE REVIEW

File: src/OpenTelemetry/Metrics/Aggregators/DoubleCounterSumAggregator.cs [hunk 1]
```diff
@@ -0,0 +1,64 @@
...
+    /// </summary>
+    public class DoubleCounterSumAggregator : Aggregator<double>
+    {
+        private double sum;
+        private double checkPoint;
+
+        public override void Checkpoint()
+        {
+            // checkpoints the current running sum into checkpoint, and starts counting again.
+            this.checkPoint = Interlocked.Exchange(ref Unsafe.As<double, double>(ref this.sum), 0.0);
```
Comment (cijothomas) [inline src/OpenTelemetry/Metrics/Aggregators/DoubleCounterSumAggregator.cs:35]: Do we still need Unsafe.As usage?
this.checkPoint = Interlocked.Exchange(ref this.sum, 0.0)

File: src/OpenTelemetry/Metrics/Aggregators/Int64CounterSumAggregator.cs [hunk 2]
```diff
@@ -0,0 +1,58 @@
...
+    /// </summary>
+    public class Int64CounterSumAggregator : Aggregator<long>
+    {
+        private long sum;
+        private long checkPoint;
+
+        public override void Checkpoint()
+        {
+            // checkpoints the current running sum into checkpoint, and starts counting again.
+                this.checkPoint = Interlocked.Exchange(ref Unsafe.As<long, long>(ref this.sum), 0);
```
Comment (cijothomas) [inline src/OpenTelemetry/Metrics/Aggregators/Int64CounterSumAggregator.cs:35]: no need of Unsafe?

File: src/OpenTelemetry/Metrics/Aggregators/DoubleCounterSumAggregator.cs [hunk 3]
```diff
@@ -0,0 +1,64 @@
...
+        public override void Update(double value)
+        {
+            // Adds value to the running total in a thread safe manner.
+            double initialTotal, computedTotal;
+            do
+            {
+                initialTotal = this.sum;
+                computedTotal = initialTotal + value;
+            }
+            while (initialTotal != Interlocked.CompareExchange(ref Unsafe.As<double, double>(ref this.sum), computedTotal, initialTotal));
```
Comment (MikeGoldsmith) [inline src/OpenTelemetry/Metrics/Aggregators/DoubleCounterSumAggregator.cs:61]: Same as above, not sure we need `Unsafe.As`

File: src/OpenTelemetry/Metrics/Aggregators/DoubleCounterSumAggregator.cs [hunk 4]
```diff
@@ -0,0 +1,64 @@
...
+                Timestamp = DateTime.UtcNow,
+            };
+        }
+
+        public override AggregationType GetAggregationType()
+        {
+            return AggregationType.DoubleSum;
+        }
+
+        public override void Update(double value)
```
Comment (MikeGoldsmith) [inline src/OpenTelemetry/Metrics/Aggregators/DoubleCounterSumAggregator.cs:52]: Do we have a `lock` here too?

File: src/OpenTelemetry/Metrics/Aggregators/Int64CounterSumAggregator.cs [hunk 5]
```diff
@@ -0,0 +1,58 @@
...
+
+        public override AggregationType GetAggregationType()
+        {
+            return AggregationType.LongSum;
+        }
+
+        public override void Update(long value)
+        {
+            // Adds value to the running total in a thread safe manner.
+            Interlocked.Add(ref Unsafe.As<long, long>(ref this.sum), value);
```
Comment (MikeGoldsmith) [inline src/OpenTelemetry/Metrics/Aggregators/Int64CounterSumAggregator.cs:55]: `Unsafe.As`

File: src/OpenTelemetry/Metrics/Aggregators/DoubleCounterSumAggregator.cs (hunk 4 above)
Comment (cijothomas) [reply src/OpenTelemetry/Metrics/Aggregators/DoubleCounterSumAggregator.cs:52]: No lock here. This keeps Sum in lock-free way.

File: src/OpenTelemetry/Metrics/Aggregators/DoubleLastValueAggregator.cs [hunk 6]
```diff
@@ -22,47 +22,31 @@ namespace OpenTelemetry.Metrics.Aggregators
...
-            {
-                throw new Exception("Invalid Type");
-            }
-        }
+        private double value;
+        private double checkpoint;
 
         public override void Checkpoint()
         {
             this.checkpoint = this.value;
```
Comment (pjanotti) [inline src/OpenTelemetry/Metrics/Aggregators/DoubleLastValueAggregator.cs:32]: Not from your change @alexeyzimarev but could you please change here and on Update to use `Interlocked.Exchange` so #585 is also fixed.

File: src/OpenTelemetry/Metrics/Aggregators/Int64LastValueAggregator.cs [hunk 7]
```diff
@@ -0,0 +1,54 @@
...
+    /// Simple aggregator that only keeps the last value.
+    /// </summary>
+    public class Int64LastValueAggregator : Aggregator<long>
+    {
+        private long value;
+        private long checkpoint;
+
+        public override void Checkpoint()
+        {
+            this.checkpoint = this.value;
```
Comment (pjanotti) [inline src/OpenTelemetry/Metrics/Aggregators/Int64LastValueAggregator.cs:32]: Could you please change here and on `Update` to use Interlocked.Exchange so #585 is also fixed.

File: src/OpenTelemetry/Metrics/Aggregators/DoubleCounterSumAggregator.cs (hunk 1 above)
Comment (alexeyzimarev) [reply src/OpenTelemetry/Metrics/Aggregators/DoubleCounterSumAggregator.cs]: Done

**Comment by alexeyzimarev** [comment]
> Thanks.
> We should remove this package from the .csproj - System.Runtime.CompilerServices.Unsafe
> 
> There may be conflicts between this and PR 547.

@MikeGoldsmith removed the `Unsafe` package reference

File: src/OpenTelemetry/Metrics/Aggregators/Int64LastValueAggregator.cs (hunk 7 above)
Comment (alexeyzimarev) [reply src/OpenTelemetry/Metrics/Aggregators/Int64LastValueAggregator.cs]: Done

File: src/OpenTelemetry/Metrics/Aggregators/DoubleLastValueAggregator.cs (hunk 6 above)
Comment (alexeyzimarev) [reply src/OpenTelemetry/Metrics/Aggregators/DoubleLastValueAggregator.cs]: Done

File: src/OpenTelemetry/Metrics/Aggregators/Int64CounterSumAggregator.cs (hunk 5 above)
Comment (alexeyzimarev) [reply src/OpenTelemetry/Metrics/Aggregators/Int64CounterSumAggregator.cs]: Done

File: src/OpenTelemetry/Metrics/Aggregators/DoubleCounterSumAggregator.cs (hunk 3 above)
Comment (alexeyzimarev) [reply src/OpenTelemetry/Metrics/Aggregators/DoubleCounterSumAggregator.cs]: You suggest adding a lock around code in `Update`?

**Comment by alexeyzimarev** [comment]
> There may be conflicts between this and PR 547.

@cijothomas I added all the code from #547 to my PR.

**Comment by cijothomas** [comment]
@alexeyzimarev Can you resolve the one remaining conflict as well!

**Pull request merged**